util_for_test.go

```diff
@@ -0,0 +1,329 @@
+package surrealdb
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	rawslog "log/slog"
+	"net"
+	"os/exec"
+	"reflect"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/surrealdb/surrealdb.go/pkg/conn/gorilla"
+	"github.com/surrealdb/surrealdb.go/pkg/logger/slog"
+)
+
+var (
+	testHasSurrealCLI     bool
+	checkSurrealCLIOnce   sync.Once
+	DelayAfterServerStart time.Duration = 300 * time.Millisecond
+	DelayBeforeServerExit time.Duration = 300 * time.Millisecond
+)
+
+func Test_NewTestSurrealDB_Simple(t *testing.T) {
+	// Simple illustration of how NewTestSurrealDB can be used.
+
+	// Tests can run in parallel.
+	t.Parallel()
+
+	// NewTestSurrealDB returns endpoint (random open port used), DB instance
+	// (which is wrapped in DBForTest to provide extra methods), and a func to
+	// shutdown the SurrealDB server.
+	endpoint, db, close := NewTestSurrealDB(t)
+	defer close()
+	_ = endpoint // endpoint could be useful for some low level testing.
+
+	// DBForTest.Prepare is to run any SurrealQL for test prep. This would fail
+	// early if the provided string results in an error from SurrealDB server.
+	db.Prepare(t, "CREATE user:x SET name = 'x'")
+
+	// More interactions can simply use db instance from here on.
+}
+
+func Test_TestSurrealDBParallelSimple(t *testing.T) {
+	t.Parallel()
+	instanceCount := 20
+	for i := 0; i < instanceCount; i++ {
+		t.Run(fmt.Sprintf("parallel__%d", i), func(t *testing.T) {
+			t.Parallel()
+
+			_, db, close := NewTestSurrealDB(t)
+			defer close()
+
+			// Ensure that CREATE statement does not run against the same
+			// database (which would fail).
+			db.Prepare(t, "CREATE user:x SET name = 'x';")
+		})
+	}
+}
+
+func Test_TestSurrealDBParalleSelect(t *testing.T) {
+	t.Parallel()
+	instanceCount := 20
+	for i := 0; i < instanceCount; i++ {
+		i := i
+		t.Run(fmt.Sprintf("parallel__%d", i), func(t *testing.T) {
+			t.Parallel()
+			_, db, close := NewTestSurrealDB(t)
+			defer close()
+
+			// Ensure that CREATE statement does not run against the same
+			// database (which would fail).
+			db.Prepare(t, fmt.Sprintf("CREATE user:x SET name = 'x', index = %d;", i))
+
+			// Query the saved data.
+			data, err := db.Select("user")
+			if err != nil {
+				t.Fatalf("failed to select: %v", err)
+			}
+			jsonBytes, err := json.Marshal(data)
+			if err != nil {
+				t.Fatalf("failed to marshal to json: %v", err)
+			}
+			// Note the Index field, which would change based on the for loop.
+			type dataholder struct {
+				ID    string `json:"id,omitempty"`
+				Name  string `json:"name,omitempty"`
+				Index int    `json:"index,omitempty"`
+			}
+			d := []dataholder{}
+			err = json.Unmarshal(jsonBytes, &d)
+			if err != nil {
+				t.Fatalf("failed to unmarshal data: %v", err)
+			}
+			// Simply check all fields match as expected.
+			want := dataholder{ID: "user:x", Name: "x", Index: i}
+			if d[0].ID != want.ID || d[0].Name != want.Name || d[0].Index != want.Index {
```

Comment:
This was a really crude check, and only wrote it so that it would be easy to reason about. Usually these checks are written with something like `if want != got { /* something like t.Error(err) */ }`, and just stuck with the same handling. I'm happy to update this anyhow 👍 (Again, I just chose to go without testify for the test cases in this file.)

One caveat is how the error report shows the `want` vs `d` rather than checking specific item such as `d[0].Name` in my implementation. It is possible that the server returns some extra data (such as `d[1]`, and thus it would be helpful to write everything rather than just asserting on a single field.